### PR TITLE
Avoid unnecessary TLB flushing during user-kernel switching 

### PR DESCRIPTION
**Pull request by anbo225**
Fix https://github.com/asterinas/asterinas/issues/799

### CODE REVIEW

**Comment by tatetian** [comment]
I think the change proposed in this PR is correct.

As a further optimization, `switch_to_task` should do nothing if the given `next_task` is equal to the `current_task`. In this case, the context switch, including the call to `activate`, should be avoided. This optimization is simple enough and can be piggybacks on this PR.

**Comment by anbo225** [comment]
> I think the change proposed in this PR is correct.
> 
> As a further optimization, `switch_to_task` should do nothing if the given `next_task` is equal to the `current_task`. In this case, the context switch, including the call to `activate`, should be avoided. This optimization is simple enough and can be piggybacked on this PR.

Thank you for your advice. I have a question: under what circumstances next_task might be equivalent to current_task? Additionally, if next_task and current_task are part of the same process, would it be possible to forgo refreshing the CR3 register? If so, do I need to implement this optimization on this PR?

**Comment by tatetian** [comment]
> I have a question: under what circumstances next_task might be equivalent to current_task? 

When there is only one single runnable task. This situation is common for microbenchmarks.

> Additionally, if next_task and current_task are part of the same process, would it be possible to forgo refreshing the CR3 register? If so, do I need to implement this optimization on this PR?

I found that you have done this optimization 👍

**Comment by tatetian** [comment]
@lrh2000 #821 has a commit that shares mostly the same optimization idea as this PR. But I think this PR executes a slightly better. So let's discuss it here. The major difference is that this PR chooses not to do `activate_kernel_vm`. This improves efficiency and does not harm correctness since Asterinas does not allow memory access via raw pointers. So keeping the user-space portion of the page table of the last task won't affect correctness or safety.

**Comment by lrh2000** [comment]
> @lrh2000 #821 has a commit that shares mostly the same optimization idea as this PR. But I think this PR executes a slightly better. So let's discuss it here. The major difference is that this PR chooses not to do `activate_kernel_vm`. This improves efficiency and does not harm correctness since Asterinas does not allow memory access via raw pointers. So keeping the user-space portion of the page table of the last task won't affect correctness or safety.

@tatetian  In fact, I completely agree with this point. However, when the previous user task exits and its `VmSpace` is dropped, its page table can still be loaded because the page table switch does not occur if the next task is a kernel task. This leads to serious soundness problems. So we need to fix this problem, perhaps by storing a cloned `VmSpace` in `PROCESSOR`?

**Comment by anbo225** [comment]
> > @lrh2000 #821 has a commit that shares mostly the same optimization idea as this PR. But I think this PR executes a slightly better. So let's discuss it here. The major difference is that this PR chooses not to do `activate_kernel_vm`. This improves efficiency and does not harm correctness since Asterinas does not allow memory access via raw pointers. So keeping the user-space portion of the page table of the last task won't affect correctness or safety.
> 
> @tatetian In fact, I completely agree with this point. However, when the previous user task exits and its `VmSpace` is dropped, its page table can still be loaded because the page table switch does not occur if the next task is a kernel task. This leads to serious soundness problems. So we need to fix this problem, perhaps by storing a cloned `VmSpace` in `PROCESSOR`?

Thank you for pointing out this issue. I will fix it.

**Comment by tatetian** [comment]
> > @lrh2000 #821 has a commit that shares mostly the same optimization idea as this PR. But I think this PR executes a slightly better. So let's discuss it here. The major difference is that this PR chooses not to do `activate_kernel_vm`. This improves efficiency and does not harm correctness since Asterinas does not allow memory access via raw pointers. So keeping the user-space portion of the page table of the last task won't affect correctness or safety.
> 
> @tatetian In fact, I completely agree with this point. However, when the previous user task exits and its `VmSpace` is dropped, its page table can still be loaded because the page table switch does not occur if the next task is a kernel task. This leads to serious soundness problems. So we need to fix this problem, perhaps by storing a cloned `VmSpace` in `PROCESSOR`?

Storing a cloned `VmSpace` in `PROCESSOR` does not sound right to me because the memory safety of page tables should rely on memory management code, not the scheduling-related code. Relying on the latter makes the code harder to understand and maintain.

I propose to increase the reference counter of the root page table frame by the new `PageTable` under review.

```rust
impl<'a, M: PageTableMode, E: PageTableEntryTrait, C: PagingConstsTrait> PageTable<M, E, C>
where
    [(); nr_ptes_per_node::<C>()]:,
    [(); C::NR_LEVELS]:,
{
    pub(crate) unsafe fn activate_unchecked(&self) {
        // Ensure that the root PageTableFrame always outlives the root page translation table given to the CPU
        {
            type RawPageTableFrame = SpinLock<PageTableFrame>;
            cpu_local! {
                static PAGE_TABLE_ROOT: AtomicPtr<RawPageTableFrame> = AtomicPtr::new(0);
            }

            let new_root_frame_ptr = Arc::into_raw(self.root_frame.clone());
            let old_root_frame_ptr = PAGE_TABLE_ROOT.swap(new_root_frame_ptr, Relaxed);
            if !old_root_frame_ptr.is_null() {
                unsafe { drop(Arc::from_raw(old_root_frame_ptr)); }
            }
        }
        
        activate_page_table(self.root_paddr(), CachePolicy::Writeback);
    }
}
```

File: framework/aster-frame/src/task/processor.rs
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:128]: I think the current approach is acceptable. However, I would like to ask, why not do this optimization at `VmSpace::activate`? We can load the register pointing to the page table root (`cr3` on x86), check if the page table root is the same as the `VmSpace` we're trying to activate, and only trigger the page table reload if the answer is negative.

My feeling is that the overhead of doing the check in `switch_to_task()` or doing the check in `VmSpace::activate` is comparable. However, the latter approach can additionally avoid page table reloads in the case where the context switches look like `user task A -> kernel task B -> user task A`.

Changing the semantics of `VmSpace::activate` to "activate if not already activated" makes a lot of sense to me, and if we really want to force-activate a `VmSpace` (to flush TLB?) we can always have another method.

Comment (junyang-zh) [reply framework/aster-frame/src/task/processor.rs:128]: Your suggestion makes sense. And someone can do it after #713 is merged. Or perhaps I can immediately do it now.

File: framework/aster-frame/src/task/processor.rs
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:155]: The `context_switch` function must never return when the previous task exits. So `current_task` and `next_task` must be dropped before calling `context_switch` to avoid resource leakage.

Ignoring this comment is pretty much fine, as I will do when I rebase #821.

Actually, I've added comments in #821 to emphasize this point. Have a look at https://github.com/lrh2000/asterinas/commit/fe84db307f138828e06631c70e4cfedcae2e3dd0#diff-31b4cc1d8756834466e8c9d51c750d62acae54b484bd20669e4891245ddccff1L132-R146

File: framework/aster-frame/src/task/processor.rs
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:145]: I was just wondering if the code would be clearer if we changed the type of `should_switch_mm` to an `Option<&VmSpace>`. For example, we can write:
```rust
    let vm_to_switch = next_task
        .user_space()
        .filter(|next_user_space| {
            !current_task
                .and_then(|task| task.user_space())
                .is_some_and(|cur_user_space| Arc::ptr_eq(cur_userspace, next_user_space))
        })
        .map(|next_user_space| next_user_space.vm_space());

    // ...

    if let Some(vm_space) = vm_to_switch {
        vm_space.activate();
    }
```

But it seems that the current code in this PR is clear enough, so the clarity is not really improved too much. So feel free to ignore this comment if you don't think doing so is better.

**Review by lrh2000 — APPROVED** [review]
I have no other comments for this PR. Changes in this PR already look good to me, some issues will be fixed in either #821 or #713. As those issues are not introduced in this PR, not fixing them in this PR is pretty okay.

File: framework/aster-frame/src/task/processor.rs
Comment (tatetian) [reply framework/aster-frame/src/task/processor.rs:128]: Agree. Putting the two optimizations in two places makes sense.

File: framework/aster-frame/src/task/processor.rs
Comment (anbo225) [reply framework/aster-frame/src/task/processor.rs:145]: I prefer your coding style and update accordingly.

File: framework/aster-frame/src/task/processor.rs
Comment (anbo225) [reply framework/aster-frame/src/task/processor.rs:148]: Agreed, it would be more suitable to address this issue in #821.

File: framework/aster-frame/src/task/processor.rs
Comment (anbo225) [reply framework/aster-frame/src/task/processor.rs:128]: Agree

File: framework/aster-frame/src/task/processor.rs
Comment (tatetian) [inline framework/aster-frame/src/task/processor.rs:131]: Rename `vm_to_switch` to `next_vm_space` as the latter is more consistent with other variable names (like `next_task` and `vm_space.activate`)

**Comment by junyang-zh** [comment]
#713 Already make the `VmSpace` API "activate if not activated". So checking here would have less conditions maybe.

**Comment by tatetian** [comment]
@anbo225 Need to remove one of the two optimizations from this PR as @junyang-zh has added it in the new page table. We will merge this PR after the new table PR is merged.

File: framework/aster-frame/src/task/processor.rs
Comment (junyang-zh) [reply framework/aster-frame/src/task/processor.rs:128]: Done

**Comment by anbo225** [comment]
> @anbo225 Need to remove one of the two optimizations from this PR as @junyang-zh has added it in the new page table. We will merge this PR after the new table PR is merged.

Since #713  make the VmSpace API "activate if not activated".   I think the following code is enough.  ``

**Comment by tatetian** [comment]
> > @anbo225 Need to remove one of the two optimizations from this PR as @junyang-zh has added it in the new page table. We will merge this PR after the new table PR is merged.
> 
> Done

Sorry. @junyang-zh has rolled back the optimization in the new page table PR because it causes system call tests to fail for unknown reasons. The PR is already heavy with changes so I chose to accept the PR without the optimization.

**Comment by junyang-zh** [comment]
Update: CI fails in #829 because by simply checking CR3 in `VmSpace::activate()` the kernel fails to invalidate TLB entries that are toxic.

```rust
/// Activate the page table if it is not currently activated in this CPU.
    pub(crate) fn activate(&self) {
        let current_pt = crate::arch::mm::current_page_table_paddr();
        if current_pt == self.pt.root_paddr() {
            return;
        }
        // Safety: The usermode page table is safe to activate since the kernel
        // mappings are shared.
        unsafe {
            self.pt.activate_unchecked();
        }
    }
```

Further investigations are needed. Maybe I need more hands on this. May @anbo225 check the the most suitable places that is needed to invalidate TLB?

**Comment by tatetian** [comment]
> the kernel fails to invalidate TLB entries that are toxic.

@junyang-zh I suspect that the new page table implementation is missing some necessary TLB flush. What do you think?

**Comment by anbo225** [comment]
> Update: CI fails in #829 because by simply checking CR3 in `VmSpace::activate()` the kernel fails to invalidate TLB entries that are toxic.
> 
> ```rust
> /// Activate the page table if it is not currently activated in this CPU.
>     pub(crate) fn activate(&self) {
>         let current_pt = crate::arch::mm::current_page_table_paddr();
>         if current_pt == self.pt.root_paddr() {
>             return;
>         }
>         // Safety: The usermode page table is safe to activate since the kernel
>         // mappings are shared.
>         unsafe {
>             self.pt.activate_unchecked();
>         }
>     }
> ```
> 
> Further investigations are needed. Maybe I need more hands on this. May @anbo225 check the the most suitable places that is needed to invalidate TLB in such a scenario?

After I rebase your code,  we need flush tlb when user invokes a system call 

> Update: CI fails in #829 because by simply checking CR3 in `VmSpace::activate()` the kernel fails to invalidate TLB entries that are toxic.
> 
> ```rust
> /// Activate the page table if it is not currently activated in this CPU.
>     pub(crate) fn activate(&self) {
>         let current_pt = crate::arch::mm::current_page_table_paddr();
>         if current_pt == self.pt.root_paddr() {
>             return;
>         }
>         // Safety: The usermode page table is safe to activate since the kernel
>         // mappings are shared.
>         unsafe {
>             self.pt.activate_unchecked();
>         }
>     }
> ```
> 
> Further investigations are needed. Maybe I need more hands on this. May @anbo225 check the the most suitable places that is needed to invalidate TLB in such a scenario?

File: framework/aster-frame/src/user.rs
Comment (tatetian) [reply framework/aster-frame/src/user.rs:143]: But why?

Comment (anbo225) [reply framework/aster-frame/src/user.rs:143]: For the gVisor syscall test, we cannot use sed to locate the correct blocklist. I suspect that the #713 (https://github.com/asterinas/asterinas/pull/713) is missing some necessary code to invalidate TLB.  Further investigations are needed. 
89c6623ce8a780c05bb56de00a6c06b

**Comment by lrh2000** [comment]
See https://github.com/asterinas/asterinas/pull/862

File: framework/aster-frame/src/task/processor.rs
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:128]: Actually, `VmSpace::activate` is no longer unsafe after #713 (which is quite reasonable), so maybe move it out of the unsafe block?

Comment (anbo225) [reply framework/aster-frame/src/task/processor.rs]: Done

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**
